File: backend/src/main/java/django/Hashers.java
package django;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.hash.Hashing;
import com.google.common.io.BaseEncoding;

import javax.annotation.Nullable;
import java.nio.charset.StandardCharsets;
import java.util.List;

import static com.google.common.primitives.Bytes.indexOf;

public final class Hashers {
    private Hashers() {}

    public static final PasswordHasher PBKDF2 = new PBKDF2PasswordHasher();
    public static final PasswordHasher SHA1 = new SHA1PasswordHasher();
    public static final List<PasswordHasher> HASHERS = ImmutableList.of(PBKDF2, SHA1);
    private static final PasswordHasher PREFERRED_HASHER = HASHERS.get(0);

    public enum CheckPasswordResult {
        CORRECT,
        CORRECT_BUT_MUST_UPDATE,
        INCORRECT,
    }

    public static CheckPasswordResult checkPassword(char[] password, String encoded) {
        return checkPassword(password, encoded, PREFERRED_HASHER);
    }

    public static CheckPasswordResult checkPassword(char[] password, String encoded, PasswordHasher preferred) {
        PasswordHasher hasher = identifyHasher(encoded);
        boolean hasherChanged = !hasher.algorithm().equals(preferred.algorithm());
        boolean mustUpdate = hasherChanged || preferred.mustUpdate(encoded);
        boolean isCorrect = hasher.verify(password, encoded);

        // If the hasher didn't change (we don't protect against enumeration if it
        // does) and the password should get updated, try to close the timing gap
        // between the work factor of the current encoded password and the default
        // work factor.
        if (!isCorrect && !hasherChanged && mustUpdate) {
            hasher.hardenRuntime(password, encoded);
        }

        if (isCorrect && mustUpdate) {
            return CheckPasswordResult.CORRECT_BUT_MUST_UPDATE;
        }

        return isCorrect ? CheckPasswordResult.CORRECT : CheckPasswordResult.INCORRECT;
    }

    private static PasswordHasher identifyHasher(String encoded) {
        int index = encoded.indexOf('$');
        if (index == -1) {
            throw new IllegalArgumentException();
        }
        String algorithm = encoded.substring(0, index);
        return HASHERS.stream().filter(it -> it.algorithm().equals(algorithm)).findFirst().get();
    }

    public static String makePassword(char[] password, @Nullable byte[] salt) {
        return makePassword(password, salt, PREFERRED_HASHER);
    }

    /**
     * Turn a plain-text password into a hash for database storage
     *
     * @param password
     * @param salt
     * @param hasher
     * @return
     */
    public static String makePassword(char[] password, @Nullable byte[] salt, PasswordHasher hasher) {
        if (salt == null) {
            salt = hasher.salt();
        }

        return hasher.encode(password, salt);
    }

    private static abstract class BasePasswordHasher implements PasswordHasher {
        @Override
        public byte[] salt() {
            return Crypto.getRandomString();
        }

        @Override
        public boolean mustUpdate(String encoded) {
            return false;
        }
    }

    private static class PBKDF2PasswordHasher extends BasePasswordHasher {
        public static final String ALGORITHM = "pbkdf2_sha256";
        private final int iterations;

        public PBKDF2PasswordHasher() {
            this(100000);
        }

        public PBKDF2PasswordHasher(int iterations) {
            this.iterations = iterations;
        }

        @Override
        public String algorithm() {
            return ALGORITHM;
        }

        @Override
        public String encode(char[] password, byte[] salt) {
            return encode(password, salt, iterations);
        }

        private String encode(char[] password, byte[] salt, int iterations) {
            Preconditions.checkNotNull(password);
            Preconditions.checkNotNull(salt);
            Preconditions.checkArgument(indexOf(salt, (byte) '$') == -1, "'$' not in salt");
            byte[] hashBytes = Crypto.pbkdf2(password, salt, iterations);
            String hash = BaseEncoding.base64().encode(hashBytes);
            return ALGORITHM + "$" + iterations + "$" + new String(salt) + "$" + hash;
        }

        @Override
        public boolean verify(char[] password, String encoded) {
            String[] parts = encoded.split("\\$", 4);
            String algorithm = parts[0];
            int iterations = Integer.parseInt(parts[1]);
            byte[] salt = parts[2].getBytes();
            Preconditions.checkArgument(algorithm.equals(ALGORITHM));
            String encoded2 = encode(password, salt, iterations);
            return Crypto.constantTimeCompare(encoded, encoded2);
        }

        @Override
        public boolean mustUpdate(String encoded) {
            String[] parts = encoded.split("\\$", 4);
            int iterations = Integer.parseInt(parts[1]);
            return iterations != this.iterations;
        }

        @Override
        public void hardenRuntime(char[] password, String encoded) {
            String[] parts = encoded.split("\\$", 4);
            int iterations = Integer.parseInt(parts[1]);
            byte[] salt = parts[2].getBytes();
            int extraIterations = this.iterations - iterations;
            if (extraIterations > 0) {
                encode(password, salt, extraIterations);
            }
        }
    }

    private static class SHA1PasswordHasher extends BasePasswordHasher {
        public static final String ALGORITHM = "sha1";

        @Override
        public String algorithm() {
            return ALGORITHM;
        }

        @Override
        public String encode(char[] password, byte[] salt) {
            Preconditions.checkNotNull(password);
            Preconditions.checkNotNull(salt);
            Preconditions.checkArgument(indexOf(salt, (byte) '$') == -1, "'$' not in salt");
            byte[] hashBytes = Hashing.sha1().newHasher()
                    .putBytes(salt)
                    .putBytes(new String(password).getBytes(StandardCharsets.UTF_8))
                    .hash().asBytes();
            String hash = BaseEncoding.base32Hex().encode(hashBytes);
            return ALGORITHM + "$" + new String(salt) + "$" + hash;
        }

        @Override
        public boolean verify(char[] password, String encoded) {
            String[] parts = encoded.split("\\$", 4);
            String algorithm = parts[0];
            byte[] salt = parts[1].getBytes();
            Preconditions.checkArgument(algorithm.equals(ALGORITHM));
            String encoded2 = encode(password, salt);
            return Crypto.constantTimeCompare(encoded, encoded2);
        }

        @Override
        public void hardenRuntime(char[] password, String encoded) {
        }
    }
}

File: backend-ts/src/auth/django/hashers.ts
import { PasswordHasher } from "./password_hasher";
import { getRandomString } from "./crypto";
import { createHash, pbkdf2, timingSafeEqual } from "crypto";
import { promisify } from "util";
import base32Encode from 'base32-encode'

export enum CheckPasswordResult {
  CORRECT,
  CORRECT_BUT_MUST_UPDATE,
  INCORRECT,
}

export function checkPassword(password: string, encoded: string): Promise<CheckPasswordResult> {
  return checkPasswordWithHasher(password, encoded, PREFERRED_HASHER)
}

export async function checkPasswordWithHasher(password: string, encoded: string, preferred: PasswordHasher): Promise<CheckPasswordResult> {
  const hasher = identifyHasher(encoded)
  const hasherChanged = hasher.algorithm() !== preferred.algorithm()
  const mustUpdate = hasherChanged || preferred.mustUpdate(encoded)
  const isCorrect = await hasher.verify(password, encoded)

  // If the hasher didn't change (we don't protect against enumeration if it
  // does) and the password should get updated, try to close the timing gap
  // between the work factor of the current encoded password and the default
  // work factor.
  if (!isCorrect && !hasherChanged && mustUpdate) {
    hasher.hardenRuntime(password, encoded)
  }

  if (isCorrect && mustUpdate) {
    return CheckPasswordResult.CORRECT_BUT_MUST_UPDATE
  }

  return isCorrect ? CheckPasswordResult.CORRECT : CheckPasswordResult.INCORRECT
}

function identifyHasher(encoded: string): PasswordHasher {
  const index = encoded.indexOf('$')
  if (index === -1)
    throw new Error('invalid format')
  const algorithm = encoded.substring(0, index)
  const hasher = HASHERS.find(it => it.algorithm() === algorithm)
  if (!hasher)
    throw new Error(`invalid algorithm: ${algorithm}`)
  return hasher
}

/**
 * Turn a plain-text password into a hash for database storage
 *
 * @param password
 * @param salt
 * @param hasher
 * @return
 */
export async function makePassword(
  password: string,
  salt: Buffer | null,
  hasher: PasswordHasher = PREFERRED_HASHER
): Promise<string> {
  if (!salt) salt = hasher.salt()
  return hasher.encode(password, salt)
}

abstract class BasePasswordHasher implements PasswordHasher {
  abstract algorithm(): string
  abstract verify(password: string, encoded: string): Promise<boolean>
  abstract encode(password: string, salt: Buffer): Promise<string>
  abstract hardenRuntime(password: string, encoded: string): Promise<void>
  
  salt(): Buffer {
    return getRandomString()
  }

  mustUpdate(encoded: string): boolean {
    return false
  }
}

class PBKDF2PasswordHasher extends BasePasswordHasher {
  static ALGORITHM = 'pbkdf2_sha256'
  static pbkdf2 = promisify(pbkdf2)

  constructor(private iterations: number = 100000) {
    super()
  }

  algorithm() {
    return PBKDF2PasswordHasher.ALGORITHM
  }

  async encode(
    password: string,
    salt: Buffer,
    iterations: number = this.iterations
  ): Promise<string> {
    if (salt.includes('$'))
      throw new Error("salt should not contain '$'")
    const hashBytes = await PBKDF2PasswordHasher.pbkdf2(password, salt, iterations, 256 / 8, 'sha256')
    const hash = hashBytes.toString('base64')
    return PBKDF2PasswordHasher.ALGORITHM + '$' + iterations + '$' + salt.toString('ascii') + '$' + hash
  }

  async verify(password: string, encoded: string): Promise<boolean> {
    const [algorithm, iterationsStr, saltStr, ] = encoded.split('$', 4)
    const iterations = Number(iterationsStr)
    const salt = Buffer.from(saltStr)
    if (algorithm !== PBKDF2PasswordHasher.ALGORITHM)
      throw new Error('invalid algorithm')
    const encoded2 = await this.encode(password, salt, iterations)
    return timingSafeEqual(Buffer.from(encoded), Buffer.from(encoded2))
  }
  
  mustUpdate(encoded: string): boolean {
    const parts = encoded.split('$', 4)
    const iterations = Number(parts[1])
    return iterations !== this.iterations
  }

  async hardenRuntime(password: string, encoded: string): Promise<void> {
    const parts = encoded.split('$', 4)
    const iterations = Number(parts[1])
    const salt = Buffer.from(parts[2])
    const extraIterations = this.iterations - iterations
    if (extraIterations > 0)
      await this.encode(password, salt, extraIterations)
  }
}

class SHA1PasswordHasher extends BasePasswordHasher {
  static ALGORITHM = 'sha1'

  algorithm() {
    return SHA1PasswordHasher.ALGORITHM
  }

  async encode(
    password: string,
    salt: Buffer
  ): Promise<string> {
    if (salt.includes('$'))
      throw new Error("salt should not contain '$'")
    const hashBytes = createHash('sha1')
      .update(salt)
      .update(password, 'utf8')
      .digest()
    const hash = base32Encode(hashBytes, 'RFC4648-HEX')
    return SHA1PasswordHasher.ALGORITHM + '$' + salt.toString('ascii') + '$' + hash
  }

  async verify(password: string, encoded: string): Promise<boolean> {
    const [algorithm, saltStr, ] = encoded.split('$', 4)
    const salt = Buffer.from(saltStr)
    if (algorithm !== SHA1PasswordHasher.ALGORITHM)
      throw new Error('invalid algorithm')
    const encoded2 = await this.encode(password, salt)
    return timingSafeEqual(Buffer.from(encoded), Buffer.from(encoded2))
  }

  async hardenRuntime(password: string, encoded: string): Promise<void> {
  }
}

export const PBKDF2: PasswordHasher = new PBKDF2PasswordHasher()
export const SHA1: PasswordHasher = new SHA1PasswordHasher()
export const HASHERS: PasswordHasher[] = [PBKDF2, SHA1]
export const PREFERRED_HASHER = HASHERS[0]